Mathlib/NumberTheory/Cyclotomic/Three.lean

```diff
@@ -109,3 +113,81 @@ theorem eq_one_or_neg_one_of_unit_of_congruent (hcong : ∃ n : ℤ, λ ^ 2 ∣
     have : (hζ.pow_of_coprime 2 (by decide)).toInteger = hζ.toInteger ^ 2 := by ext; simp
     simp only [this, PNat.val_ofNat, Nat.cast_ofNat, mul_neg, Int.cast_neg, ← neg_add, ←
       sub_eq_iff_eq_add.1 hx, Units.val_neg, val_pow_eq_pow_val, IsUnit.unit_spec, neg_neg]
+
+variable (x : 𝓞 K)
+
+/-- Let `(x : 𝓞 K)`. Then we have that `λ` divides one amongst `x`, `x - 1` and `x + 1`. -/
+lemma lambda_dvd_or_dvd_sub_one_or_dvd_add_one : λ ∣ x ∨ λ ∣ x - 1 ∨ λ ∣ x + 1 := by
+  classical
+  have := hζ.finite_quotient_toInteger_sub_one (by decide)
+  let _ := Fintype.ofFinite (𝓞 K ⧸ Ideal.span {λ})
+  let _ : Ring (𝓞 K ⧸ Ideal.span {λ}) := CommRing.toRing -- to speed up instance synthesis
+  let _ : AddGroup (𝓞 K ⧸ Ideal.span {λ}) := AddGroupWithOne.toAddGroup -- dito
+  have := Finset.mem_univ (Ideal.Quotient.mk (Ideal.span {λ}) x)
+  have h3 : Fintype.card (𝓞 K ⧸ Ideal.span {λ}) = 3 := by
+    rw [← Nat.card_eq_fintype_card, hζ.card_quotient_toInteger_sub_one (by decide),
+      hζ.norm_toInteger_sub_one_of_prime_ne_two' (by decide)]
+    simp only [PNat.val_ofNat, Nat.cast_ofNat, Int.reduceAbs]
+  rw [Finset.univ_of_card_eq_three h3] at this
+  simp only [Finset.mem_insert, Finset.mem_singleton] at this
+  rcases this with h | h | h
+  · left
+    exact Ideal.mem_span_singleton.1 <| Ideal.Quotient.eq_zero_iff_mem.1 h
+  · right; left
+    refine Ideal.mem_span_singleton.1 <| Ideal.Quotient.eq_zero_iff_mem.1 ?_
+    rw [RingHom.map_sub, h, RingHom.map_one, sub_self]
+  · right; right
+    refine Ideal.mem_span_singleton.1 <| Ideal.Quotient.eq_zero_iff_mem.1 ?_
+    rw [RingHom.map_add, h, RingHom.map_one, add_left_neg]
+
+/-- We have that `η ^ 2 + η + 1 = 0`. -/
+lemma eta_sq_add_eta_add_one : (η : 𝓞 K) ^ 2 + η + 1 = 0 := by
+  rw [eta_sq]
+  ring
+
+/-- We have that `x ^ 3 - 1 = (x - 1) * (x - η) * (x - η ^ 2)`. -/
+lemma cube_sub_one_eq_mul : x ^ 3 - 1 = (x - 1) * (x - η) * (x - η ^ 2) := by
+  symm
+  calc _ = x ^ 3 - x ^ 2 * (η ^ 2 + η + 1) + x * (η ^ 2 + η + η ^ 3) - η ^ 3 := by ring
+  _ = x ^ 3 - x ^ 2 * (η ^ 2 + η + 1) + x * (η ^ 2 + η + 1) - 1 := by
+    simp [show hζ.toInteger ^ 3 = 1 from hζ.toInteger_isPrimitiveRoot.pow_eq_one]
+  _ = x ^ 3 - 1 := by rw [eta_sq_add_eta_add_one hζ]; ring
+
+/-- We have that `λ` divides `x * (x - 1) * (x - (η + 1))`. -/
+lemma lambda_dvd_mul_sub_one_mul_sub_eta_add_one : λ ∣ x * (x - 1) * (x - (η + 1)) := by
+  rcases lambda_dvd_or_dvd_sub_one_or_dvd_add_one hζ x with h | h | h
+  · exact dvd_mul_of_dvd_left (dvd_mul_of_dvd_left h _) _
+  · exact dvd_mul_of_dvd_left (dvd_mul_of_dvd_right h _) _
+  · refine dvd_mul_of_dvd_right ?_ _
+    rw [show x - (η + 1) = x + 1 - (η - 1 + 3) by ring]
+    exact dvd_sub h <| dvd_add dvd_rfl hζ.toInteger_sub_one_dvd_prime'
```

Comment:
More note to self than anything to do in this PR: I feel like this should be `aesop`able...